File: day_04/day4b.js
const fs = require('fs');

const inputText = fs.readFileSync('./day_04/day4_input.txt', 'utf-8');
const rows = inputText.split('\n');

const multiplier = Array.from({ length: rows.length }, () => 1);
rows.forEach((row, r) => {
    cutCard = row.slice(8);
    const parts = cutCard.split('|');

    const convertToArray = (str) => {
        return str.trim().split(/\s+/).map(num => parseInt(num, 10));
    };
    const winning = convertToArray(parts[0]);
    const ours = convertToArray(parts[1]);
    var num_wins = 0
    ours.forEach(num => {
        if (winning.includes(num)) {
            num_wins += 1;
        };
    });
    for (let i = 0; i < num_wins; i++) {
        multiplier[r+i+1] += multiplier[r]
    }
})
var sum = multiplier.reduce((acc, curr) => acc + curr, 0);
console.log(sum)

File: day_04/day4b.py
import math

with open("day_04/day4_input.txt", 'r') as file:
    characters_array = [line.strip() for line in file]

asum = 0
multiplier = [1] * len(characters_array)
for r, row in enumerate(characters_array):
    split_array = row.split('|')
    winning = [int(num) for num in split_array[0].split() if num.isdigit()]
    ours = [int(num) for num in split_array[1].split() if num.isdigit()]
    num_wins = 0
    for num in ours:
        if num in winning:
            num_wins += 1
    for i in range(num_wins):
        multiplier[i+r+1] += multiplier[r]
    print(multiplier)
print(sum(multiplier))

